homeassistant/components/esphome/entry_data.py

```diff
@@ -125,18 +133,65 @@ async def async_update_static_infos(
         async_dispatcher_send(hass, signal, infos)
 
     @callback
-    def async_update_state(self, hass: HomeAssistant, state: EntityState) -> None:
+    def register_state_type(
+        self, state_type: type[EntityState], component_key: str
+    ) -> None:
+        """Register a state type."""
+        self.entity_state_type_to_component_key[state_type] = component_key
+        self.entity_state_type_name_to_component_key[
+            state_type.__name__
+        ] = component_key
+
+    @callback
+    def component_key_for_state(self, state: EntityState) -> str:
+        """Return the component key for a state."""
+        state_class_name = state.__class__.__name__
+
+        # We try to avoid doing a linear search for the
+        # the component key by looking up the class name
+        # in the mapping. If it's not there, we fall back
+        # to a linear search for the component key by doing
+        # an instance search over entity_state_type_to_component_key.
+        if state_class_name in self.entity_state_type_name_to_component_key:
+            return self.entity_state_type_name_to_component_key[state_class_name]
+
+        for (
+            state_type,
+            component_key,
+        ) in self.entity_state_type_to_component_key.items():
+            if isinstance(state, state_type):
+                return component_key
+
+        raise ValueError(f"{state} is not a registered state_type")
+
+    @callback
+    def async_subscribe_state_update(
+        self,
+        component_key: str,
+        state_key: int,
+        entity_callback: Callable[[], None],
+    ) -> Callable[[], None]:
+        """Subscribe to state updates."""
+
+        def _unsubscribe() -> None:
+            self.state_subscriptions.pop((component_key, state_key))
+
+        self.state_subscriptions[(component_key, state_key)] = entity_callback
+        return _unsubscribe
```

Comment:
I mimicked a cut down version of what we do with DataUpdateCoordinator here since there can only be one key per entity https://github.com/home-assistant/core/blob/24d2d6212135bea2e1a2e8d8f35d967001124811/homeassistant/components/esphome/__init__.py#L582

https://github.com/home-assistant/core/blob/24d2d6212135bea2e1a2e8d8f35d967001124811/homeassistant/helpers/update_coordinator.py#L85